br/pkg/utils/iter/iter.go

```diff
@@ -32,6 +32,12 @@ type TryNextor[T any] interface {
 	TryNext(ctx context.Context) IterResult[T]
 }
 
+// TryNextEnumor is the general interface for "impure" iterators:
+// which may trigger some error or block the caller when advancing.
+type TryNextEnumor[T any] interface {
```

Comment:
I think perhaps this interface can be replaced with something like:
```go
type Indexed[T any] struct {
  Index int
  Item T
}

type TryNextEnumor[T] = TryNextor[Indexed[T]]

type withIndex[T] struct {
  inner tryNextor[T]
  index int
}

func (wr *withIndex[T]) TryNext(ctx context.Context) Result[Indexed[T]] {
  r := wr.inner.TryNext(ctx)
  if r.Done || r.Err != nil {
    return r
  }
  res := Emit(Indexed{
    Index: wr.index,
    Item: r.Item
  })
  wr.index ++
  return res
}

func Enumerate[T](it TryNextor[T]) TryNextEnumor[T] {
  return withIndex[T] { inner: it, index: 0 }
}
```

So we don't need to reimplement the `FilterOutEnum` or something, just call `FilterOut(Enumerate(it), pred)`.